Mage/Archivist/Archivist.cs

```diff
@@ -54,8 +76,166 @@ public class Archivist : EventEmitter<JObject> {
+		if (cacheValue == null) {
+			cacheValue = new VaultValue(topic, index);
+			_cache.Add(cacheKeyName, cacheValue);
+		}
+
+		JObject rawValue = (JObject)info["value"];
+		if (rawValue != null) {
+			ValueSet(topic, index, rawValue["data"], (string)rawValue["mediaType"], (int?)info["expirationTime"]);
+		} else {
+			ValueDel(topic, index);
+		}
+	}
+
+
+	// Return cache dictionary
+	public Dictionary<string, VaultValue> GetCache() {
+		return _cache;
+	}
+
+
+	// Clear out the cache entirely
+	public void ClearCache() {
+		_cache = new Dictionary<string, VaultValue>();
+	}
+
+
+	// Remove a vault value from the cache by it's topic and index
+	public void DeleteCacheItem(string topic, JObject index) {
+		DeleteCacheItem(GetCacheKey(topic, index));
+	}
+
+
+	// Remove a vault value from the cache by it's cache key name
+	public void DeleteCacheItem(string cacheKeyName) {
+		if (!_cache.ContainsKey(cacheKeyName)) {
+			return;
+		}
+
+		_cache.Remove(cacheKeyName);
+	}
+
+
+	////////////////////////////////////////////
+	//        Vault Value Manipulation        //
+	////////////////////////////////////////////
+	private void ValueSet(string topic, JObject index, JToken data, string mediaType, int? expirationTime) {
+		// Try and get cache value. If it exists delete existing value
+		// in preparation for set. Otherwise create a new vault value.
+		string cacheKeyName = GetCacheKey(topic, index);
+		VaultValue cacheValue = GetCacheValue(cacheKeyName);
+		if (cacheValue == null) {
+			cacheValue = new VaultValue(topic, index);
+			_cache.Add(cacheKeyName, cacheValue);
+		} else {
+			cacheValue.Del();
+		}
+
+		// Set data to vault value
+		cacheValue.SetData(mediaType, data);
+		cacheValue.Touch(expirationTime);
+
+		// Emit set event
+		this.emit(topic + ":set", cacheValue);
+	}
+
+	private void ValueAdd(string topic, JObject index, JToken data, string mediaType, int? expirationTime) {
+		// Check if value already exists
+		string cacheKeyName = GetCacheKey(topic, index);
+		VaultValue cacheValue = GetCacheValue(cacheKeyName);
+		if (cacheValue != null) {
+			logger.error("Could not add value (already exists): " + cacheKeyName);
+			return;
+		}
+
+		// Create new vault value
+		cacheValue = new VaultValue(topic, index);
+		_cache.Add(cacheKeyName, cacheValue);
+
+		// Set data to vault value
+		cacheValue.SetData(mediaType, data);
+		cacheValue.Touch(expirationTime);
+		
+		// Emit add event
+		this.emit(topic + ":add", cacheValue);
+	}
+
+	private void ValueDel(string topic, JObject index) {
+		// Check if value already exists
+		string cacheKeyName = GetCacheKey(topic, index);
+		VaultValue cacheValue = GetCacheValue(cacheKeyName);
+		if (cacheValue == null) {
+			logger.error("Could not delete value (doesn't exist): " + cacheKeyName);
+			return;
+		}
+		
+		// Do delete
+		cacheValue.Del();
+		
+		// Emit touch event
+		this.emit(topic + ":del", cacheValue);
+	}
+
+	private void ValueTouch(string topic, JObject index, int? expirationTime) {
```

Comment:
the backend will not provide an expiration time (undefined) if there is no TTL on the data